clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java

```
@@ -60,221 +55,126 @@
+
+                        if (version < TOP_LEVEL_ERROR_AND_NULL_TOPICS_MIN_VERSION && group.errorCode() != Errors.NONE.code()) {
+                            // Versions prior to version 2 do not support a top level error. Therefore,
+                            // we put it at the partition level.
+                            newPartition = new OffsetFetchResponsePartition()
+                                .setPartitionIndex(partition.partitionIndex())
+                                .setErrorCode(group.errorCode())
+                                .setCommittedOffset(INVALID_OFFSET)
+                                .setMetadata(NO_METADATA)
+                                .setCommittedLeaderEpoch(NO_PARTITION_LEADER_EPOCH);
+                        } else {
+                            newPartition = new OffsetFetchResponsePartition()
+                                .setPartitionIndex(partition.partitionIndex())
+                                .setErrorCode(partition.errorCode())
+                                .setCommittedOffset(partition.committedOffset())
+                                .setMetadata(partition.metadata())
+                                .setCommittedLeaderEpoch(partition.committedLeaderEpoch());
+                        }
+
+                        newTopic.partitions().add(newPartition);
+                    });
+                });
             }
 
-            OffsetFetchResponseGroup group = groups.get(0);
-            data.setErrorCode(group.errorCode());
-            error = Errors.forCode(group.errorCode());
-
-            group.topics().forEach(topic -> {
-                OffsetFetchResponseTopic newTopic = new OffsetFetchResponseTopic().setName(topic.name());
-                data.topics().add(newTopic);
-
-                topic.partitions().forEach(partition -> {
-                    OffsetFetchResponsePartition newPartition;
-
-                    if (version < 2 && group.errorCode() != Errors.NONE.code()) {
```

There is a bug here. This logic assumes that partitions are given along side a group level error. However, looking at KafkaApis, it is not the case. Hence, a group level error is not returned at all for version 1 of the API. I will file a Jira and address it separately.

---

https://issues.apache.org/jira/browse/KAFKA-19246